Content promotion & UI redesign

forgottosave opened the pull request:
PR for [Issue #11](https://github.com/AP-Sensing/ostree-tui/issues/11)

# Content Promotion
This PR brings **Content Promotion** to the **OSTree TUI**.

## Basics
To make it fit into the UI, I reworked the basic layout to use **Tabs** on the left halve of the screen (where previously the branch filter and commit info was). This also allows for an easy addition of future features in new tabs, while creating a neat and modern look for the UI. In addition to that, the *Footer* is now able to display notifications to the user. 

## Implementation
The *Manager* is now split into 4 parts, as it is now only responsible for putting the 3 new tabs together (instead of being responsible for all that happens within this view). The 3 tabs each got their own class, to store data and logical components:
- the *CommitInfoManager*,
- the *BranchBoxManager*,
- and the *ContentPromotionManager*.

## ToDos
- [x] Restructure UI layout -> Tabs
- [x] Add content promotion to UI
- [x] Implement content promotion in `cpplibostree`
- [x] Code Cleanup
- [ ] Make content promotion direct `libostree` access (not CLI)

*@COM8  [suggestions](https://github.com/AP-Sensing/ostree-tui/pull/12#issuecomment-2144581861):*
- [x] Add Tooltips
- [x] Improve promotion-tab navigation
- [x] Fix UI component size & flex
- [x] Rethink boxed layout

forgottosave committed:
cpplibostree setup

forgottosave committed:
promotion manager setup

forgottosave committed:
footer content-promotion-setup

forgottosave committed:
Implement content promotion ref selection

forgottosave committed:
Implement content promotion in cpplibostree (CLI format)

implementation has to be replaced by the proper libostree access in future

forgottosave committed:
Implement content promotion UI

forgottosave committed:
Implement footer notifications

forgottosave committed:
partial code cleanup refactor

after UI implementation of the tabs and the content promotion UI, this all needs to be cleaned up (away from main, into separate Managers for each content. My brain is absolutely fried right now...

forgottosave committed:
finish interchangeable view refactor

the call to the actual content promotion remains to be implemented

forgottosave committed:
remove unused exec

forgottosave committed:
Minor UI modifications

forgottosave committed:
Change FTXUI version 4.0.0 -> 5.0.0

forgottosave committed:
Connect UI to OSTree promotion

until now, the call to teh OSTree promotion was not done yet. Now the content promotion works as intended, although the CLI-access implementation in cppostree should probably be replaced with a proper libostree-access

forgottosave committed:
Fix promotion subject

forgottosave committed:
Update FTXUI to most recent version 5.0.0 -> main

forgottosave committed:
Change Shortcuts

Ctrl instead of Alt shortcuts would probably be better and more intuitive for the user, but don't work properly yet in FTXUI (bug?)

forgottosave committed:
Change content promotion layout

forgottosave committed:
Remove add-metadata-strings section

Potentially gets added in the commit-info view, to also be modifyable

COM8 commented:
Things I noticed:

### New Branch Box
When I maximize my window and I have a lot of branches, there is a lot of white space not used and instead I have to scroll through my branches.

| Current| Suggested |
| ----- | ----- |
| ![Screenshot from 2024-06-03 10-02-06](https://github.com/AP-Sensing/ostree-tui/assets/11741404/b55e72f1-e4ef-4a19-b8af-1b324aca326f) | ![image](https://github.com/AP-Sensing/ostree-tui/assets/11741404/7c9768a4-ba49-4ad3-b1f4-02e9c29eead3) |

I suggest extending the new branch list to fill the complete remaining space but perhaps with a maximum hight so it does not occupy all the vspace.

### Commit

![image](https://github.com/AP-Sensing/ostree-tui/assets/11741404/ed32dcec-baa3-46bb-8277-ef779b744a29)

I do not understand what you mean with "Commit...". Perhaps it would be better to either move the tab list to the next line, or remove the "Commit..." label/text completely since it does not do anything at all.

### Tooltips?
Add tooltips if possible to all options or perhaps, if there is enough vspace display a single help string at the bottom of the terminal. Perhaps even with a small ℹ️ identifier at the beginning?

Example:
![image](https://github.com/AP-Sensing/ostree-tui/assets/11741404/d3b714d9-004d-43ea-a780-e1cb6f1003b3)


### Navigation
How do I skip the branch selection. Do I really have to scroll through all branches to get to the "Apply" button.

### Info

Display commit metadata inside the "Info" tab and if possible let me modify it. Add a warning that this requires resigning the commit.

### C.-Checksum

Better call it just Checksum?

### Boxed Layout

In "Promote" everything is inside a "Box". In Info only the Subject. Why? Also the alignment inside the info tab is not correct. Content is missing the indentation.

forgottosave committed:
Remove Commit... label

forgottosave committed:
Leave promotion-branch-selection on Return

This doesn't force the user to scroll through all branches, just to get to the apply button

forgottosave committed:
Renname C.-Checksum -> Checksum and signatures -> Signatures with bulletpoints

forgottosave committed:
Add Manager Documentation

forgottosave committed:
Add Tool-Tips to Promotion Tab

forgottosave committed:
Fix promotion view element-sizes

Calculates window size and the resulting element distributions. The tooptips only get shown if the window is sufficiently tall.

forgottosave commented:
@COM8 thank you for this review. I added your suggestions into the ToDo list, but created a [new issue](https://github.com/AP-Sensing/ostree-tui/issues/14) for your **Info** suggestions, as I would count it as a new feature (even though very useful for the content promotion workflow).

forgottosave committed:
Redesign box-layout

'Boxed' elements now clearly separate user-input from only displayed text in the promotion view

forgottosave committed:
Code Cleanup

forgottosave committed:
Add option to hide tooltips

COM8 commented:
Perfect. I will look into it tomorrow again :)

forgottosave commented:
the remaining ToDo, `Make content promotion direct libostree access (not CLI)` will be moved to another issue as well. As far as this PR goes, it is ready to merge :)
